Add `days_left` field

TheAndrewJackson opened the pull request:
# Purpose
Add dynamic `days_left` field that is generated based off of the current `due_date` value of a privacy request.

# Changes
- Add new property to `PrivacyRequest` model
- Add new tests and update existing tests to work with new update

# Checklist
- [x] Update [`CHANGELOG.md`](https://github.com/ethyca/fidesops/blob/main/CHANGELOG.md) file
  - [x] Merge in main so the most recent `CHANGELOG.md` file is being appended to
  - [x] Add description within the `Unreleased` section in an appropriate category. Add a new category from the list at the top of the file if the needed one isn't already there.
  - [x] Add a link to this PR at the end of the description with the PR number as the text. example: [#1](https://github.com/ethyca/fidesops/pull/1)
- [ ] Applicable documentation updated (guides, quickstart, postman collections, tutorial, fidesdemo, [database diagram](https://github.com/ethyca/fidesops/blob/main/docs/fidesops/docs/development/update_erd_diagram.md).
- If docs updated (select one):
  - [ ] documentation complete, or draft/outline provided (tag docs-team to complete/review on this branch)
  - [ ] documentation issue created (tag docs-team to complete issue separately)
- [x] Good unit test/integration test coverage
- [ ] This PR contains a DB migration. If checked, the reviewer should confirm with the author that the [down_revision correctly references the previous migration](https://ethyca.github.io/fidesops/development/contributing_details/#alembic-migrations) before merging
- [ ] The `Run Unsafe PR Checks` label has been applied, and checks have passed, if this PR touches any external services

# Ticket

Fixes #1148

sanders41 commented on src/fidesops/ops/models/privacy_request.py:
```diff
@@ -138,6 +138,14 @@ def generate_request_callback_jwe(webhook: PolicyPreWebhook) -> str:
     return generate_jwe(json.dumps(jwe.dict()), config.security.app_encryption_key)
 
 
+def get_days_left(self: PrivacyRequest) -> Union[int, None]:
+    if self.due_date is None:
+        return None
+
+    delta = self.due_date.date() - datetime.now(timezone.utc).date()
+    return delta.days
```
I would rename `self` here. `self` makes it look like it is part of the class but it isn't so I initially thought it wasn't indented correctly.

You could also slightly simplify the utc date with `utcnow`.

```suggestion
def get_days_left(privacy_request: PrivacyRequest) -> Union[int, None]:
    if privacy_request.due_date is None:
        return None

    delta = privacy_request.due_date.date() - datetime.utcnow().date()
    return delta.days
```

sanders41 commented on src/fidesops/ops/models/privacy_request.py:
```diff
@@ -212,6 +220,7 @@ class PrivacyRequest(Base):  # pylint: disable=R0904
     paused_at = Column(DateTime(timezone=True), nullable=True)
     identity_verified_at = Column(DateTime(timezone=True), nullable=True)
     due_date = Column(DateTime(timezone=True), nullable=True)
+    days_left = property(get_days_left)
```
I am guessing it has something to do with sqlalchemy being unhappy, but is there a reason this was defined as a property this way instead of:

```py
@property
def days_left(self)...
```

TheAndrewJackson replied:
There isn't a specific reason. I'll try this out and see if it works. The `property` function/decorator is newer to me so I didn't realize I could do it that way.

sanders41 commented on tests/ops/api/v1/endpoints/test_privacy_request_endpoints.py:
```diff
@@ -1322,6 +1334,41 @@ def test_get_failed_request_resume_info_from_email_send(
         }
         assert data["resume_endpoint"] == f"/privacy-request/{privacy_request.id}/retry"
 
+    @pytest.mark.parametrize(
```
Making use of the parametrization, I like it!

TheAndrewJackson replied:
Thanks! I've been looking for opportunities to use it

TheAndrewJackson commented:
@sanders41 I updated the `days_left` field to use the `property` decorator and moved the function body inside the model. Since the function is inside the model now I left the parameter as self. If that doesn't make sense for the `property` decorator I can push up a quick update changing it

sanders41 commented:
The property update looks good. I'll approve and merge as soon as CI finishes.

sanders41 commented:
@TheAndrewJackson one thing I just realized, this updates a database model. Do we need a migration?

TheAndrewJackson commented:
@sanders41 AFAIK we don't need a migration. We decided to calculate this property at the api level because it's a value that changes every day and there could be many changes if the org has a lot of privacy requests. 

If at a later date we decided to move this property into the DB and have an automated job that update all the due dates we would need a migration.
